src/main/java/com/answerdigital/answerking/service/TagService.java

```diff
@@ -0,0 +1,54 @@
+package com.answerdigital.answerking.service;
+
+import com.answerdigital.answerking.mapper.TagMapper;
+import com.answerdigital.answerking.model.Product;
+import com.answerdigital.answerking.model.Tag;
+import com.answerdigital.answerking.repository.TagRepository;
+import com.answerdigital.answerking.request.TagRequest;
+import com.answerdigital.answerking.response.TagResponse;
+import org.mapstruct.factory.Mappers;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Set;
+
+import static com.answerdigital.answerking.exception.util.GlobalErrorMessage.TAGS_ALREADY_EXIST;
+import static com.answerdigital.answerking.exception.util.GlobalErrorMessage.getCustomException;
+
+@Service
+public class TagService {
+    private final TagRepository tagRepository;
+
+    private final TagMapper tagMapper = Mappers.getMapper(TagMapper.class);
+
+    private final ProductService productService;
+
+    @Autowired
+    public TagService(final TagRepository tagRepository, final ProductService productService) {
+        this.tagRepository = tagRepository;
+        this.productService = productService;
+    }
+
+    public TagResponse addTag(final TagRequest tagRequest) {
+        if (tagRepository.existsByName(tagRequest.name())) {
+            throw getCustomException(TAGS_ALREADY_EXIST, tagRequest.name());
+        }
+
+        final List<Product> products = productService.findAllProductsInListOfIds(tagRequest.productIds());
```

Comment:
This needs to be able to return a 404 if one or more productIds do not exist. We're imagining an approach where we'd compare the IDs in the request with the IDs of the returned objects, and we can then get the difference of the two lists and return that list in an error message to the front end.

Reply:
UPDATE - We'll make this a separate ticket as it'll require API spec changes.

Reply:
❌ Won't fix on the basis of separating this into a separate ticket.